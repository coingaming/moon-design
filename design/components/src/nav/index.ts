export * from './nav';
export * from './nav-link';
