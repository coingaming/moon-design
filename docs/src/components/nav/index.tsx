/** @jsx jsx */
import { StaticQuery, graphql } from 'gatsby';
import { jsx, CSSObject } from '@emotion/core';
import { colors } from '@heathmont/sportsbet-tokens';
import { spacing } from '@heathmont/sportsbet-utils';
import { PagesData } from '../../types';
import { transformSections } from './transform';
import { Divider } from './divider';
import { Logo } from './logo';
import { Menu, MenuList, MenuItem } from './menu';

jsx;

const sidebar = graphql`
  query {
    pages: allMdx {
      edges {
        node {
          frontmatter {
            name
            menu
            route
          }
        }
      }
    }
    config: allDocsJson {
      edges {
        node {
          designSections
        }
      }
    }
  }
`;

const nav: CSSObject = {
  backgroundColor: colors.neutral[90],
  padding: spacing(),
};

export const Nav = () => (
  <StaticQuery
    query={sidebar}
    render={(data: PagesData) => (
      <nav css={nav}>
        <Logo />
        <Divider />
        <MenuList>
          <MenuItem route="/" name="Introduction" section />
          {/* TODO: Investigate IDE type issue highlighting */}
          <Menu items={transformSections(data)} />
        </MenuList>
      </nav>
    )}
  />
);
