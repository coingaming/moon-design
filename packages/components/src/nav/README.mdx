---
name: Nav
menu: Components
route: /components/nav
---

import { Nav } from '@heathmont/sportsbet-components';
import { PropsTable } from 'docs';

# Nav

A menu of items for users to move between sections of the application.

By default, the nav will provide an accessible skip link, and overflow with
horizontal scrolling.

Each `NavLink` must be wrapped in a `<NavItem>`.

```jsx react-live
<Nav>
  <NavItem>
    <NavLink href="#1" active>
      Item 1
    </NavLink>
  </NavItem>
  <NavItem>
    <NavLink href="#2">Item 2</NavLink>
  </NavItem>
  <NavItem>
    <NavLink href="#3">Item 3</NavLink>
  </NavItem>
  <NavItem>
    <NavLink href="#4">Item 4</NavLink>
  </NavItem>
  <NavItem>
    <NavLink href="#5">Item 5</NavLink>
  </NavItem>
</Nav>
```

## Anatomy

### Link

By default, the `NavLink` renders as an [anchor tag](https://developer.mozilla.org/en-US/docs/Web/HTML/Element/a),
accepting the standard attributes.

```jsx react-live
<NavLink href="#1">Item 1</NavLink>
```

However, it sometimes makes more semantic sense to mark up the link as a button.
We also might want to use a custom `Link` component for a specific router.

For these reasons, `NavLink` can accept a custom `element`.

```jsx react-live
<NavLink element="button" onClick={() => alert('Hello')}>
  Item 1
</NavLink>
```
