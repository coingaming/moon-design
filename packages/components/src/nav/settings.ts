import { spacing } from '@heathmont/sportsbet-utils';
import { border } from '@heathmont/sportsbet-tokens';

const underlineWidth = border.width * 2;
const underlineOffset = spacing('small');

export { underlineOffset, underlineWidth };
