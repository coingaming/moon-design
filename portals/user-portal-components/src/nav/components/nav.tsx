/** @jsx jsx */
import { jsx, CSSObject } from '@emotion/core';
import styled from '@emotion/styled';
import hideVisually from 'polished/lib/mixins/hideVisually';
import { colors, breakpoints, typography } from '@heathmont/sportsbet-tokens';
import { spacing, mq } from '@heathmont/sportsbet-utils';
import rem from 'polished/lib/helpers/rem';
import { Link } from '@heathmont/sportsbet-components';
import { closeStyles } from './hamburger/hamburger-menu';

jsx;

const navSkipLinkID = 'nav-skip';

const skipLink: CSSObject = {
  position: 'absolute',
  zIndex: 1,
  padding: `0 ${spacing('small')}`,
  top: '50%',
  left: spacing('small'),
  transform: 'translateY(-50%)',
  color: colors.neutral[10],
  backgroundColor: colors.neutral[90],
  '&:not(:focus)': {
    ...hideVisually(),
  },
};

const Nav = styled.nav(() => [
  {
    width: '100%',
    backgroundColor: colors.neutral[90],
    overflow: 'hidden',
  },
]);

const navGrid: CSSObject = {
  fontWeight: typography.fontWeight.semibold,
  display: 'grid',
  gridTemplateRows: `${rem(64)} auto auto`,
  gridTemplateColumns: '1fr 1fr',
  gridColumnGap: spacing(),
  gridTemplateAreas: `
      "hamburger    buttons"
      "logo         logo"
      "search       search"
    `,
  marginTop: 0,
  marginLeft: 0,
  height: '100%',
  [mq(breakpoints.small)]: {
    paddingLeft: spacing('large'),
    paddingRight: spacing('large'),
  },
  [mq(breakpoints.large)]: {
    gridTemplateColumns: 'fit-content(20%) auto fit-content(20%)',
    gridTemplateRows: `${rem(64)}`,
    gridTemplateAreas: `
        "hamburger    search    buttons"
      `,
  },
};

const Navigation: React.FC = ({ children }) => {
  return (
    <Nav>
      <a href={`#${navSkipLinkID}`} css={skipLink}>
        Skip to content
      </a>
      <div css={navGrid}>{children}</div>
      <span id={navSkipLinkID} />
    </Nav>
  );
};

const CloseBack = styled(Link)({
  ...closeStyles,
  marginLeft: spacing('small'),
  [mq(breakpoints.medium)]: {
    display: 'none',
  },
});

export { Navigation, CloseBack };
