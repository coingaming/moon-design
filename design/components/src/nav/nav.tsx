/** @jsx jsx */
import * as React from 'react';
import { jsx, CSSObject } from '@emotion/core';
import styled from '@emotion/styled';
import hideVisually from 'polished/lib/mixins/hideVisually';
import math from 'polished/lib/math/math';
import rem from 'polished/lib/helpers/rem';
import { colors } from '@heathmont/sportsbet-tokens';
import { listInline, listInlineItem } from '@heathmont/sportsbet-objects';
import { spacing, rhythm } from '@heathmont/sportsbet-utils';
import { underlineOffset, underlineWidth } from './settings';

/**
 * Settings
 */
const navSkipLinkID = 'nav-skip';

/**
 * Components
 */
const Nav: React.FC = ({ children }) => {
  const nav: CSSObject = {
    ...rhythm(),
    position: 'relative',
    overflowX: 'scroll',
    overflowY: 'hidden',
  };

  const skipLink: CSSObject = {
    position: 'absolute',
    zIndex: 1,
    padding: `0 ${spacing('small')}`,
    top: '50%',
    left: spacing('small'),
    transform: 'translateY(-50%)',
    color: colors.neutral[10],
    backgroundColor: colors.neutral[90],
    '&:not(:focus)': {
      ...hideVisually(),
    },
  };

  /* 1. Include the indicator as part of the item's box-model. */
  const navList: CSSObject = {
    ...listInline,
    display: 'block',
    whiteSpace: 'nowrap',
    marginTop: 0,
    paddingBottom: math(
      `${underlineOffset} + ${rem(underlineWidth)}`
    ) /* [1] */,
  };

  return (
    <nav css={nav}>
      <a href={`#${navSkipLinkID}`} css={skipLink}>
        Skip to content
      </a>
      <ul css={navList}>{children}</ul>
      <span id={navSkipLinkID} />
    </nav>
  );
};

const NavItem = styled.li([
  listInlineItem,
  {
    /* Add spacing to each adjacent item */
    '& + &': {
      marginLeft: spacing(),
    },
  },
]);

/**
 * Exports
 */
export { Nav, NavItem };
