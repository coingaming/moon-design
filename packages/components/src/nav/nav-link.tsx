/** @jsx jsx */
import * as React from 'react';
import { jsx, css } from '@emotion/core';
import { animation, colors } from '@heathmont/sportsbet-tokens';
import { spacing } from '@heathmont/sportsbet-utils';
import { underlineWidth } from './settings';

/**
 * Types
 *
 * 1. String or React Component.
 * 2. Allows us to pass in any props associated to the `element`.
 */
type NavLinkProps = {
  active?: boolean;
  element?: any /* [1] */;
  [prop: string]: any /* [2] */;
};

/**
 * Component
 */
const NavLink: React.FC<NavLinkProps> = ({ active, element = 'a', ...props }) =>
  jsx(element, {
    ...props,
    'aria-current': active ? 'page' : undefined,
    css: css({
      position: 'relative',
      padding: 0,
      margin: 0,
      fontSize: 'inherit',
      fontFamily: 'inherit',
      backgroundColor: 'transparent',
      color: colors.neutral[20],
      transition: `color ${animation.speed.default}s ease`,
      textDecoration: 'none',
      border: 0,
      appearance: 'none',
      '&:after': {
        content: '""',
        position: 'absolute',
        height: `${underlineWidth}px`,
        width: spacing('medium'),
        left: 0,
        right: 0,
        bottom: `-${spacing('small')}`,
        margin: '0 auto',
        backgroundColor: colors.brand,
        transform: 'scaleX(0)',
        transition: `transform ${animation.speed.default}s ease`,
      },
      '&:hover, &:focus, &:active, &[aria-current=page]': {
        color: colors.neutral[10],
        cursor: 'pointer',
        '&:after': {
          transform: 'scaleX(1)',
        },
      },
    }),
  });

export { NavLink, NavLinkProps };
