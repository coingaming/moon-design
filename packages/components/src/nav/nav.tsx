/** @jsx jsx */
import * as React from 'react';
import { jsx, css } from '@emotion/core';
import styled from '@emotion/styled';
import hideVisually from 'polished/lib/mixins/hideVisually';
import math from 'polished/lib/math/math';
import rem from 'polished/lib/helpers/rem';
import { colors } from '@heathmont/sportsbet-tokens';
import { listInline, listInlineItem } from '@heathmont/sportsbet-objects';
import { spacing } from '@heathmont/sportsbet-utils';
import { underlineOffset, underlineWidth } from './settings';

/**
 * Settings
 */
const navSkipLinkID = 'nav-skip';

/**
 * Components
 */
const Nav: React.FC = ({ children }) => {
  const nav = css({
    position: 'relative',
    overflowX: 'scroll',
    overflowY: 'hidden',
  });

  const skipLink = css({
    position: 'absolute',
    zIndex: 1,
    padding: `0 ${spacing('small')}`,
    top: '50%',
    left: spacing('small'),
    transform: 'translateY(-50%)',
    color: colors.neutral[10],
    backgroundColor: colors.neutral[90],
    '&:not(:focus)': {
      ...hideVisually(),
    },
  });

  /**
   * 1. Start the vertical rhythm from the bottom of the underline indicator.
   */
  const navList = css([
    listInline,
    {
      display: 'block',
      whiteSpace: 'nowrap',
      marginBottom: math(
        `${underlineOffset} + ${spacing('default')} + ${rem(underlineWidth)}`
      ) /* [1] */,
    },
  ]);

  return (
    <nav css={nav}>
      <a href={`#${navSkipLinkID}`} css={skipLink}>
        Skip to content
      </a>
      <ul css={navList}>{children}</ul>
      <span id={navSkipLinkID} />
    </nav>
  );
};

const NavItem = styled.li([
  listInlineItem,
  {
    /* Add spacing to each adjacent item */
    '& + &': {
      marginLeft: spacing(),
    },
  },
]);

/**
 * Exports
 */
export { Nav, NavItem };
