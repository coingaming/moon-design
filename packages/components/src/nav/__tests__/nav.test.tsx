import * as React from 'react';
import { create } from 'react-test-renderer';

import { Nav, NavItem } from '../nav';

describe('Nav', () => {
  test('renders correctly', () => {
    const navLink = create(
      <Nav>
        {[1, 2, 3, 4].map(i => (
          <NavItem key={i}>
            Item
            {i}
          </NavItem>
        ))}
      </Nav>
    );

    expect(navLink).toMatchSnapshot();
  });
});
