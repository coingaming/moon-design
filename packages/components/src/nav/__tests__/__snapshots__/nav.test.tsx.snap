// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Nav renders correctly 1`] = `
.emotion-6 {
  margin-top: 1rem;
  position: relative;
  overflow-x: scroll;
  overflow-y: hidden;
}

.emotion-0 {
  position: absolute;
  z-index: 1;
  padding: 0 0.5rem;
  top: 50%;
  left: 0.5rem;
  -webkit-transform: translateY(-50%);
  -ms-transform: translateY(-50%);
  transform: translateY(-50%);
  color: #FFF;
  background-color: #1A212A;
}

.emotion-0:not(:focus) {
  border: 0;
  -webkit-clip: rect(0 0 0 0);
  clip: rect(0 0 0 0);
  -webkit-clip-path: inset(50%);
  clip-path: inset(50%);
  height: 1px;
  margin: -1px;
  overflow: hidden;
  padding: 0;
  position: absolute;
  white-space: nowrap;
  width: 1px;
}

.emotion-5 {
  display: block;
  margin-left: 0;
  white-space: nowrap;
  margin-top: 0;
  padding-bottom: 0.625rem;
}

.emotion-1 {
  display: inline-block;
  list-style-type: none;
}

.emotion-1:before {
  position: absolute;
  content: "\\200B";
}

.emotion-1 + .emotion-1 {
  margin-left: 1rem;
}

<nav
  className="emotion-6"
>
  <a
    className="emotion-0"
    href="#nav-skip"
  >
    Skip to content
  </a>
  <ul
    className="emotion-5"
  >
    <li
      className="emotion-1"
    >
      Item
      1
    </li>
    <li
      className="emotion-1"
    >
      Item
      2
    </li>
    <li
      className="emotion-1"
    >
      Item
      3
    </li>
    <li
      className="emotion-1"
    >
      Item
      4
    </li>
  </ul>
  <span
    id="nav-skip"
  />
</nav>
`;
